### PR TITLE
CORDA-3188: Ignore synthetic and static fields when searching for state pointers

### PR DESCRIPTION
**Pull request by shamsasari**
Fix for https://github.com/corda/corda/issues/5362

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt [hunk 1]
```diff
@@ -55,31 +62,24 @@ class StatePointerSearch(val state: ContractState) {
     }
 
     private fun handleObject(obj: Any?) {
-        if (obj == null) return
-        seenObjects.add(obj)
+        if (obj == null || !seenObjects.add(obj)) return
         when (obj) {
             is Map<*, *> -> handleMap(obj)
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt]: I would like to skip all SingletonSerializeAsToken objects too as that protects the border of our framework.

Comment (roger3cev) [reply core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt]: Fair but none of those objects _should_ ever be included in a state object!

File: core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt [hunk 2]
```diff
@@ -21,16 +23,22 @@ class StatePointerSearch(val state: ContractState) {
...
-    private val seenObjects = Collections.newSetFromMap(IdentityHashMap<Any, Boolean>()).apply { add(state) }
+    private val seenObjects = Collections.newSetFromMap(IdentityHashMap<Any, Boolean>())
 
     // Queue of fields to search.
-    private val fieldQueue = ArrayDeque<FieldWithObject>().apply { addAllFields(state) }
+    private val fieldQueue = ArrayDeque<FieldWithObject>()
+
+    init {
+        seenObjects += state
+        addAllFields(state)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt]: Is it more correct to replace both lines with `addObject(state)` ?

Comment (shamsasari) [reply core/src/main/kotlin/net/corda/core/internal/StatePointerSearch.kt]: yep

**Pull request merged**